### PR TITLE
8288043: Optimize FP to word/sub-word integral type conversion on X86 AVX2 platforms

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Hi All,

This patch extends conversion optimizations added with [JDK-8287835](https://bugs.openjdk.org/browse/JDK-8287835) to optimize following floating point to integral conversions for X86 AVX2 targets:-
 * D2I , D2S,  D2B, F2I ,  F2S,  F2B
 
In addition, it also optimizes following wide vector (64 bytes) double to integer and sub-type conversions for AVX512 targets which do not support AVX512DQ feature. 
 * D2I,  D2S, D2B
  
Following are the JMH micro performance results with and without patch.

System configuration: 40C 2S Icelake server (Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz) 

BENCHMARK | SIZE | BASELINE (ops/ms) | WITHOPT (ops/ms) | PERF GAIN FACTOR
-- | -- | -- | -- | --
VectorFPtoIntCastOperations.microDouble128ToByte128 | 1024 | 90.603 | 92.797 | 1.024215534
VectorFPtoIntCastOperations.microDouble128ToByte256 | 1024 | 81.909 | 82.3 | 1.00477359
VectorFPtoIntCastOperations.microDouble128ToByte512 | 1024 | 26.181 | 26.244 | 1.002406325
VectorFPtoIntCastOperations.microDouble128ToInteger128 | 1024 | 90.74 | 2537.958 | 27.96956138
VectorFPtoIntCastOperations.microDouble128ToInteger256 | 1024 | 81.586 | 2429.599 | 29.7796068
VectorFPtoIntCastOperations.microDouble128ToInteger512 | 1024 | 19.406 | 19.61 | 1.010512213
VectorFPtoIntCastOperations.microDouble128ToLong128 | 1024 | 91.723 | 90.754 | 0.989435583
VectorFPtoIntCastOperations.microDouble128ToShort128 | 1024 | 91.766 | 1984.577 | 21.62649565
VectorFPtoIntCastOperations.microDouble128ToShort256 | 1024 | 81.949 | 1940.599 | 23.68056962
VectorFPtoIntCastOperations.microDouble128ToShort512 | 1024 | 16.468 | 16.56 | 1.005586592
VectorFPtoIntCastOperations.microDouble256ToByte128 | 1024 | 163.331 | 3018.351 | 18.479964
VectorFPtoIntCastOperations.microDouble256ToByte256 | 1024 | 148.878 | 3082.034 | 20.70174237
VectorFPtoIntCastOperations.microDouble256ToByte512 | 1024 | 50.108 | 51.629 | 1.030354434
VectorFPtoIntCastOperations.microDouble256ToInteger128 | 1024 | 159.805 | 4619.421 | 28.90661118
VectorFPtoIntCastOperations.microDouble256ToInteger256 | 1024 | 143.876 | 4649.642 | 32.31700909
VectorFPtoIntCastOperations.microDouble256ToInteger512 | 1024 | 38.127 | 38.188 | 1.001599916
VectorFPtoIntCastOperations.microDouble256ToLong128 | 1024 | 160.322 | 162.442 | 1.013223388
VectorFPtoIntCastOperations.microDouble256ToLong256 | 1024 | 141.252 | 143.01 | 1.012445841
VectorFPtoIntCastOperations.microDouble256ToShort128 | 1024 | 157.717 | 3757.471 | 23.82413437
VectorFPtoIntCastOperations.microDouble256ToShort256 | 1024 | 143.876 | 3830.971 | 26.62689399
VectorFPtoIntCastOperations.microDouble256ToShort512 | 1024 | 32.061 | 32.911 | 1.026511962
VectorFPtoIntCastOperations.microFloat128ToByte128 | 1024 | 146.599 | 4002.967 | 27.30555461
VectorFPtoIntCastOperations.microFloat128ToByte256 | 1024 | 136.99 | 3938.799 | 28.75245638
VectorFPtoIntCastOperations.microFloat128ToByte512 | 1024 | 51.561 | 50.284 | 0.975233219
VectorFPtoIntCastOperations.microFloat128ToInteger128 | 1024 | 5933.565 | 5361.472 | 0.903583596
VectorFPtoIntCastOperations.microFloat128ToInteger256 | 1024 | 5079.564 | 5062.046 | 0.996551279
VectorFPtoIntCastOperations.microFloat128ToInteger512 | 1024 | 37.101 | 38.419 | 1.035524649
VectorFPtoIntCastOperations.microFloat128ToLong128 | 1024 | 145.863 | 145.362 | 0.99656527
VectorFPtoIntCastOperations.microFloat128ToLong256 | 1024 | 131.159 | 133.154 | 1.015210546
VectorFPtoIntCastOperations.microFloat128ToShort128 | 1024 | 145.966 | 4150.039 | 28.4315457
VectorFPtoIntCastOperations.microFloat128ToShort256 | 1024 | 134.703 | 4566.589 | 33.90116775
VectorFPtoIntCastOperations.microFloat128ToShort512 | 1024 | 31.878 | 30.867 | 0.968285338
VectorFPtoIntCastOperations.microFloat256ToByte128 | 1024 | 237.841 | 6292.051 | 26.4548627
VectorFPtoIntCastOperations.microFloat256ToByte256 | 1024 | 222.041 | 6292.748 | 28.34047766
VectorFPtoIntCastOperations.microFloat256ToByte512 | 1024 | 92.073 | 88.981 | 0.966417951
VectorFPtoIntCastOperations.microFloat256ToInteger128 | 1024 | 11471.121 | 10269.636 | 0.895260019
VectorFPtoIntCastOperations.microFloat256ToInteger256 | 1024 | 10729.816 | 10105.92 | 0.941853989
VectorFPtoIntCastOperations.microFloat256ToInteger512 | 1024 | 68.328 | 70.005 | 1.024543379
VectorFPtoIntCastOperations.microFloat256ToLong128 | 1024 | 247.101 | 248.571 | 1.005948984
VectorFPtoIntCastOperations.microFloat256ToLong256 | 1024 | 225.74 | 223.987 | 0.992234429
VectorFPtoIntCastOperations.microFloat256ToLong512 | 1024 | 76.39 | 76.187 | 0.997342584
VectorFPtoIntCastOperations.microFloat256ToShort128 | 1024 | 233.196 | 8202.179 | 35.17289748
VectorFPtoIntCastOperations.microFloat256ToShort256 | 1024 | 220.75 | 7781.073 | 35.24834881
VectorFPtoIntCastOperations.microFloat256ToShort512 | 1024 | 58.143 | 55.633 | 0.956830573

Kindly review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288043](https://bugs.openjdk.org/browse/JDK-8288043): Optimize FP to word/sub-word integral type conversion on X86 AVX2 platforms

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to [5fb99e81](https://git.openjdk.org/jdk/pull/9748/files/5fb99e815eb875a74ed2054888e9046cfb1b2a67)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9748/head:pull/9748` \
`$ git checkout pull/9748`

Update a local copy of the PR: \
`$ git checkout pull/9748` \
`$ git pull https://git.openjdk.org/jdk pull/9748/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9748`

View PR using the GUI difftool: \
`$ git pr show -t 9748`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9748.diff">https://git.openjdk.org/jdk/pull/9748.diff</a>

</details>

### CODE REVIEW

**Comment by TobiHartmann** [comment]
I can run some testing in our system once you resolved the merge conflicts.

**Comment by TobiHartmann** [comment]
Testing in our system did not show any failures but I see that there are SIGILL failures in the pre-submit testing.

**Comment by sviswa7** [comment]
Could you please enable the compiler/vectorapi/VectorFPtoIntCastTest.java for AVX2 platforms?
Currently they are only run for AVX512DQ platforms.

**Comment by jatin-bhateja** [comment]
> Could you please enable the compiler/vectorapi/VectorFPtoIntCastTest.java for AVX2 platforms? Currently they are only run for AVX512DQ platforms.

I have added missing casting cases AVX/AVX2 and AVX512 targets in existing comprehensive test for [casting](test/hotspot/jtreg/compiler/vectorapi/reshape/utils/TestCastMethods.java.)

File: src/hotspot/cpu/x86/matcher_x86.hpp
Comment (vnkozlov) [inline src/hotspot/cpu/x86/matcher_x86.hpp:193]: Add comment here and in other cases explaining numbers. Is it size of instructions or elements or something?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/matcher_x86.hpp]: Value now matches the one for RoundV[FD] IR nodes, currently, its a rudimentary heuristic based on emitted code size for complex IR nodes used by unroll policy.

Comment (vnkozlov) [reply src/hotspot/cpu/x86/matcher_x86.hpp]: Please, add this information/clarification to this method's comment at line 186.

**Review by vnkozlov — COMMENTED** [review]
Good. I will test it.

You need second review.

**Review by vnkozlov — APPROVED** [review]
My testing passed.

**Review by sviswa7 — COMMENTED** [review]
I am still going through the c2_MacroAssembler_x86.cpp changes. Hopefully early next week will finish the review.

**Comment by sviswa7** [comment]
@jatin-bhateja Rest of the changes look good to me. Mainly the vector_op_pre_select_sz_estimate() needs to be corrected.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4611]: Hi @sviswa7, assertions are part of leaf level macro assembly routine which is vector_cast_float_to_long_special_cases_evex in this case.

**Review by sviswa7 — APPROVED** [review]

**Comment by vnkozlov** [comment]
@jatin-bhateja, please merge latest JDK and I will start re-testing.

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja, please merge latest JDK and I will start re-testing.

Hi @kvn, kindly regress the changes.

**Review by vnkozlov — APPROVED** [review]
My testing passed.

**Comment by jatin-bhateja** [comment]
Thanks @sviswa7 and @vnkozlov.

**Pull request closed without merging**
